src/evict/evict_lru.c

```diff
@@ -1987,6 +1995,233 @@ __evict_walk_tree(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, u_int max_ent
+  uint64_t min_pages, uint32_t target_pages)
+{
+    WT_BTREE *btree;
+    bool give_up;
+
+    btree = S2BT(session);
+
+    /*
+     * Check whether we're finding a good ratio of candidates vs pages seen. Some workloads create
+     * "deserts" in trees where no good eviction candidates can be found. Abandon the walk if we get
+     * into that situation.
+     */
+    give_up = !__wt_cache_aggressive(session) && !WT_IS_HS(btree->dhandle) &&
+      pages_seen > min_pages &&
+      (pages_queued == 0 || (pages_seen / pages_queued) > (min_pages / target_pages));
+    if (give_up) {
+        /*
+         * Try a different walk start point next time if a walk gave up.
+         */
+        switch (btree->evict_start_type) {
+        case WT_EVICT_WALK_NEXT:
+            btree->evict_start_type = WT_EVICT_WALK_PREV;
+            break;
+        case WT_EVICT_WALK_PREV:
+            btree->evict_start_type = WT_EVICT_WALK_RAND_PREV;
+            break;
+        case WT_EVICT_WALK_RAND_PREV:
+            btree->evict_start_type = WT_EVICT_WALK_RAND_NEXT;
+            break;
+        case WT_EVICT_WALK_RAND_NEXT:
+            btree->evict_start_type = WT_EVICT_WALK_NEXT;
+            break;
+        }
+
+        /*
+         * We differentiate the reasons we gave up on this walk and increment the stats accordingly.
+         */
+        if (pages_queued == 0) {
+            WT_STAT_CONN_INCR(session, cache_eviction_walks_gave_up_no_targets);
+            WT_STAT_DSRC_INCR(session, cache_eviction_walks_gave_up_no_targets);
+        } else {
+            WT_STAT_CONN_INCR(session, cache_eviction_walks_gave_up_ratio);
+            WT_STAT_DSRC_INCR(session, cache_eviction_walks_gave_up_ratio);
+        }
+    }
+
+    return (give_up);
+}
+
+/*
+ * __evict_try_queue_page --
+ *     Check if we should queue the page for eviction. Queue it to the urgent queue or the regular
+ *     queue.
+ */
+static WT_INLINE void
+__evict_try_queue_page(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, WT_REF *ref,
+  WT_PAGE *last_parent, WT_EVICT_ENTRY *evict, bool *urgent_queuedp, bool *queuedp)
+{
+    WT_BTREE *btree;
+    WT_CACHE *cache;
+    WT_CONNECTION_IMPL *conn;
+    WT_PAGE *page;
+    bool modified, want_page;
+
+    btree = S2BT(session);
+    conn = S2C(session);
+    cache = conn->cache;
+    page = ref->page;
+    modified = __wt_page_is_modified(page);
+    *queuedp = false;
+
+    /* Don't queue dirty pages in trees during checkpoints. */
+    if (modified && WT_BTREE_SYNCING(btree)) {
+        WT_STAT_CONN_INCR(session, cache_eviction_server_skip_dirty_pages_during_checkpoint);
+        return;
+    }
+
+    /*
+     * It's possible (but unlikely) to visit a page without a read generation, if we race with the
+     * read instantiating the page. Set the page's read generation here to ensure a bug doesn't
+     * somehow leave a page without a read generation.
```

Comment:
If I'm reading this correctly: I don't think this is necessary as the __wt_atomic_load64 is defined as seq_cst.

Reply:
__wt_atomic_load is implemented with relaxed semantic. It is recently added by PM-3221 to work with TSAN.

Reply:
So it should still be part of the modification order: so I don't see a data race.
I'm confused as to where the bug would be located? 
Is it appropriate to set the value here, or alternatively capture the value into variable and default to a suitable generation if it is not set; or would that risk evicting a page without the generation set?

Reply:
I also don't know. The aim here is to keep the exact behaviour as before.

Reply:
I think here we are trying to solve a case that we race with the page read. We have read the page into memory but we haven't set the read gen in the page read code yet. Therefore, we set it here to ensure it is set and the lru sorting will work correctly on this page.

Reply:
Okay thanks for clarifying. Does no generation indicate that the page has just been read into the cache? If so does it make sense to even consider it for eviction?

Reply:
Yes, it is just read into cache. That's why we set the read gen to new, which means it is very unlikely to be evicted.

Reply:
thanks for clarifying